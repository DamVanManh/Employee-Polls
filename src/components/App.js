import { useEffect } from "react";
import { connect } from "react-redux";
import { Routes, Route } from "react-router-dom";
import Navbar from "./Navbar";
import LoginPage from "./LoginPage";
import Dashboard from "./Dashboard";
import AddQuestion from "./AddQuestion";
import Leaderboard from "./Leaderboard";
import PollQuestion from "./PollQuestion";
import PageNotFound from "./PageNotFound";
import { setQuestions } from "../actions/questions";
import { setUsers } from "../actions/users";
import { _getUsers, _getQuestions } from "../utils/_DATA";

const App = ({ user, dispatch }) => {
  useEffect(() => {
    dispatch(async (dispatch) => {
      const { users, questions } = await Promise.all([
        _getUsers(),
        _getQuestions(),
      ]).then(([users, questions]) => ({
        users,
        questions,
      }));
      dispatch(setQuestions(questions));
      dispatch(setUsers(users));
    });
  }, []);
  return (
    <div className="container row d-flex justify-content-center m-auto">
      {user === null ? (
        <LoginPage />
      ) : (
        <>
          <Navbar />
          <Routes>
            <Route exact path="/" element={<Dashboard />} />
            <Route path="/leaderboard" element={<Leaderboard />} />
            <Route path="/add" element={<AddQuestion />} />
            <Route path="/questions/:question_id" element={<PollQuestion />} />
            <Route path="/pagenotfound" exact element={<PageNotFound />} />
            <Route path="*" element={<PageNotFound />} />
          </Routes>
        </>
      )}
    </div>
  );
};

const mapStateToProps = ({ user }) => ({
  user,
});

export default connect(mapStateToProps)(App);
