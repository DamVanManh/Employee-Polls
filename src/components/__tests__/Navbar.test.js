import { createStore } from "redux";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import App from "../App";
import { users, questions } from "../../utils/mockdata";

describe("Test Nav click", () => {
  it("Should move to AddQuestion page", async () => {
    const storeMock = createStore(
      (state = { users, questions, user: "sarahedo" }) => state
    );
    jest.spyOn(storeMock, "dispatch").mockImplementation(() => {});

    const view = render(
      <Provider store={storeMock}>
        <Router>
          <App />
        </Router>
      </Provider>
    );

    const link = screen.getByTestId("new-question-btn");
    fireEvent.click(link);
    await view.findByText(/Create Your Own Poll/i);
  });
});
