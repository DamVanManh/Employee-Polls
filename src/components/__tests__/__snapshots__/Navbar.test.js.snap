// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Nav click Should log out 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body />,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Test Nav click should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container row d-flex justify-content-center m-auto"
      >
        <nav
          class="navbar navbar-expand-lg navbar-light bg-light"
        >
          <div
            class="container-fluid"
          >
            <div
              class="collapse navbar-collapse"
              id="navbarSupportedContent"
            >
              <ul
                class="navbar-nav me-auto mb-2 mb-lg-0"
              >
                <li
                  class="nav-item"
                >
                  <a
                    class="nav-link active btn fs-4"
                    data-link="/"
                  >
                    Home
                  </a>
                </li>
                <li
                  class="nav-item"
                >
                  <a
                    class="nav-link active btn fs-4"
                    data-link="/leaderboard"
                  >
                    Leaderboard
                  </a>
                </li>
                <li
                  class="nav-item"
                >
                  <a
                    class="nav-link active btn fs-4"
                    data-link="/add"
                    data-testid="new-question-btn"
                  >
                    New Question
                  </a>
                </li>
              </ul>
              <div
                class="d-flex align-items-center"
              >
                <img
                  alt="Sarah Edo"
                  class="avatar"
                  src="/images/avatar1.jpeg"
                />
                <span
                  class="pe-2"
                >
                  Sarah Edo
                </span>
                <button
                  class="btn btn-outline-secondary"
                  data-testid="log-out-btn"
                >
                  Logout
                </button>
              </div>
            </div>
          </div>
        </nav>
        <div
          class="d-flex align-items-start mt-3"
        >
          <div
            aria-orientation="vertical"
            class="nav flex-column nav-pills me-3"
            id="v-pills-tab"
            role="tablist"
          >
            <button
              aria-controls="v-pills-home"
              aria-selected="true"
              class="nav-link active btn-success"
              data-bs-target="#v-pills-home"
              data-bs-toggle="pill"
              id="v-pills-home-tab"
              role="tab"
              type="button"
            >
              Unanswered polls
            </button>
            <button
              aria-controls="v-pills-profile"
              aria-selected="false"
              class="nav-link btn-success"
              data-bs-target="#v-pills-profile"
              data-bs-toggle="pill"
              id="v-pills-profile-tab"
              role="tab"
              type="button"
            >
              Answered polls
            </button>
          </div>
          <div
            class="tab-content container-fluid"
            id="v-pills-tabContent"
          >
            <div
              aria-labelledby="v-pills-home-tab"
              class="tab-pane fade show active"
              id="v-pills-home"
              role="tabpanel"
            >
              <div
                class="card p-2 mb-2"
              >
                <div
                  class="card-header text-bg-primary"
                >
                  Unanswered polls
                </div>
                <div
                  class="card-body"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col-4"
                    >
                      <div
                        class="card mb-2"
                      >
                        <h5
                          class="card-header"
                        >
                          mtsamis
                        </h5>
                        <div
                          class="card-body"
                        >
                          <h5
                            class="card-title"
                          >
                            Mon May 01 2017
                          </h5>
                          <a
                            class="navbar-link text-decoration-none"
                            href="/questions/xj352vofupe1dqz9emx13r"
                          >
                            <div
                              class="d-grid gap-2"
                            >
                              <button
                                class="btn btn-info"
                                type="button"
                              >
                                Show
                              </button>
                            </div>
                          </a>
                        </div>
                      </div>
                    </div>
                    <div
                      class="col-4"
                    >
                      <div
                        class="card mb-2"
                      >
                        <h5
                          class="card-header"
                        >
                          tylermcginnis
                        </h5>
                        <div
                          class="card-body"
                        >
                          <h5
                            class="card-title"
                          >
                            Wed Mar 15 2017
                          </h5>
                          <a
                            class="navbar-link text-decoration-none"
                            href="/questions/vthrdm985a262al8qx3do"
                          >
                            <div
                              class="d-grid gap-2"
                            >
                              <button
                                class="btn btn-info"
                                type="button"
                              >
                                Show
                              </button>
                            </div>
                          </a>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="v-pills-profile-tab"
              class="tab-pane fade"
              id="v-pills-profile"
              role="tabpanel"
            >
              <div
                class="card p-2 container-fluid"
              >
                <div
                  class="card-header text-bg-success"
                >
                  Answered polls
                </div>
                <div
                  class="card-body"
                >
                  <div
                    class="row"
                  >
                    <div
                      class="col-4"
                    >
                      <div
                        class="card mb-2"
                      >
                        <h5
                          class="card-header"
                        >
                          sarahedo
                        </h5>
                        <div
                          class="card-body"
                        >
                          <h5
                            class="card-title"
                          >
                            Sat Mar 04 2017
                          </h5>
                          <a
                            class="navbar-link text-decoration-none"
                            href="/questions/am8ehyc8byjqgar0jgpub9"
                          >
                            <div
                              class="d-grid gap-2"
                            >
                              <button
                                class="btn btn-info"
                                type="button"
                              >
                                Show
                              </button>
                            </div>
                          </a>
                        </div>
                      </div>
                    </div>
                    <div
                      class="col-4"
                    >
                      <div
                        class="card mb-2"
                      >
                        <h5
                          class="card-header"
                        >
                          tylermcginnis
                        </h5>
                        <div
                          class="card-body"
                        >
                          <h5
                            class="card-title"
                          >
                            Sat Dec 24 2016
                          </h5>
                          <a
                            class="navbar-link text-decoration-none"
                            href="/questions/loxhs1bqm25b708cmbf3g"
                          >
                            <div
                              class="d-grid gap-2"
                            >
                              <button
                                class="btn btn-info"
                                type="button"
                              >
                                Show
                              </button>
                            </div>
                          </a>
                        </div>
                      </div>
                    </div>
                    <div
                      class="col-4"
                    >
                      <div
                        class="card mb-2"
                      >
                        <h5
                          class="card-header"
                        >
                          mtsamis
                        </h5>
                        <div
                          class="card-body"
                        >
                          <h5
                            class="card-title"
                          >
                            Thu Jul 14 2016
                          </h5>
                          <a
                            class="navbar-link text-decoration-none"
                            href="/questions/6ni6ok3ym7mf1p33lnez"
                          >
                            <div
                              class="d-grid gap-2"
                            >
                              <button
                                class="btn btn-info"
                                type="button"
                              >
                                Show
                              </button>
                            </div>
                          </a>
                        </div>
                      </div>
                    </div>
                    <div
                      class="col-4"
                    >
                      <div
                        class="card mb-2"
                      >
                        <h5
                          class="card-header"
                        >
                          sarahedo
                        </h5>
                        <div
                          class="card-body"
                        >
                          <h5
                            class="card-title"
                          >
                            Wed Jun 29 2016
                          </h5>
                          <a
                            class="navbar-link text-decoration-none"
                            href="/questions/8xf0y6ziyjabvozdd253nd"
                          >
                            <div
                              class="d-grid gap-2"
                            >
                              <button
                                class="btn btn-info"
                                type="button"
                              >
                                Show
                              </button>
                            </div>
                          </a>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container row d-flex justify-content-center m-auto"
    >
      <nav
        class="navbar navbar-expand-lg navbar-light bg-light"
      >
        <div
          class="container-fluid"
        >
          <div
            class="collapse navbar-collapse"
            id="navbarSupportedContent"
          >
            <ul
              class="navbar-nav me-auto mb-2 mb-lg-0"
            >
              <li
                class="nav-item"
              >
                <a
                  class="nav-link active btn fs-4"
                  data-link="/"
                >
                  Home
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link active btn fs-4"
                  data-link="/leaderboard"
                >
                  Leaderboard
                </a>
              </li>
              <li
                class="nav-item"
              >
                <a
                  class="nav-link active btn fs-4"
                  data-link="/add"
                  data-testid="new-question-btn"
                >
                  New Question
                </a>
              </li>
            </ul>
            <div
              class="d-flex align-items-center"
            >
              <img
                alt="Sarah Edo"
                class="avatar"
                src="/images/avatar1.jpeg"
              />
              <span
                class="pe-2"
              >
                Sarah Edo
              </span>
              <button
                class="btn btn-outline-secondary"
                data-testid="log-out-btn"
              >
                Logout
              </button>
            </div>
          </div>
        </div>
      </nav>
      <div
        class="d-flex align-items-start mt-3"
      >
        <div
          aria-orientation="vertical"
          class="nav flex-column nav-pills me-3"
          id="v-pills-tab"
          role="tablist"
        >
          <button
            aria-controls="v-pills-home"
            aria-selected="true"
            class="nav-link active btn-success"
            data-bs-target="#v-pills-home"
            data-bs-toggle="pill"
            id="v-pills-home-tab"
            role="tab"
            type="button"
          >
            Unanswered polls
          </button>
          <button
            aria-controls="v-pills-profile"
            aria-selected="false"
            class="nav-link btn-success"
            data-bs-target="#v-pills-profile"
            data-bs-toggle="pill"
            id="v-pills-profile-tab"
            role="tab"
            type="button"
          >
            Answered polls
          </button>
        </div>
        <div
          class="tab-content container-fluid"
          id="v-pills-tabContent"
        >
          <div
            aria-labelledby="v-pills-home-tab"
            class="tab-pane fade show active"
            id="v-pills-home"
            role="tabpanel"
          >
            <div
              class="card p-2 mb-2"
            >
              <div
                class="card-header text-bg-primary"
              >
                Unanswered polls
              </div>
              <div
                class="card-body"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-4"
                  >
                    <div
                      class="card mb-2"
                    >
                      <h5
                        class="card-header"
                      >
                        mtsamis
                      </h5>
                      <div
                        class="card-body"
                      >
                        <h5
                          class="card-title"
                        >
                          Mon May 01 2017
                        </h5>
                        <a
                          class="navbar-link text-decoration-none"
                          href="/questions/xj352vofupe1dqz9emx13r"
                        >
                          <div
                            class="d-grid gap-2"
                          >
                            <button
                              class="btn btn-info"
                              type="button"
                            >
                              Show
                            </button>
                          </div>
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-4"
                  >
                    <div
                      class="card mb-2"
                    >
                      <h5
                        class="card-header"
                      >
                        tylermcginnis
                      </h5>
                      <div
                        class="card-body"
                      >
                        <h5
                          class="card-title"
                        >
                          Wed Mar 15 2017
                        </h5>
                        <a
                          class="navbar-link text-decoration-none"
                          href="/questions/vthrdm985a262al8qx3do"
                        >
                          <div
                            class="d-grid gap-2"
                          >
                            <button
                              class="btn btn-info"
                              type="button"
                            >
                              Show
                            </button>
                          </div>
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            aria-labelledby="v-pills-profile-tab"
            class="tab-pane fade"
            id="v-pills-profile"
            role="tabpanel"
          >
            <div
              class="card p-2 container-fluid"
            >
              <div
                class="card-header text-bg-success"
              >
                Answered polls
              </div>
              <div
                class="card-body"
              >
                <div
                  class="row"
                >
                  <div
                    class="col-4"
                  >
                    <div
                      class="card mb-2"
                    >
                      <h5
                        class="card-header"
                      >
                        sarahedo
                      </h5>
                      <div
                        class="card-body"
                      >
                        <h5
                          class="card-title"
                        >
                          Sat Mar 04 2017
                        </h5>
                        <a
                          class="navbar-link text-decoration-none"
                          href="/questions/am8ehyc8byjqgar0jgpub9"
                        >
                          <div
                            class="d-grid gap-2"
                          >
                            <button
                              class="btn btn-info"
                              type="button"
                            >
                              Show
                            </button>
                          </div>
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-4"
                  >
                    <div
                      class="card mb-2"
                    >
                      <h5
                        class="card-header"
                      >
                        tylermcginnis
                      </h5>
                      <div
                        class="card-body"
                      >
                        <h5
                          class="card-title"
                        >
                          Sat Dec 24 2016
                        </h5>
                        <a
                          class="navbar-link text-decoration-none"
                          href="/questions/loxhs1bqm25b708cmbf3g"
                        >
                          <div
                            class="d-grid gap-2"
                          >
                            <button
                              class="btn btn-info"
                              type="button"
                            >
                              Show
                            </button>
                          </div>
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-4"
                  >
                    <div
                      class="card mb-2"
                    >
                      <h5
                        class="card-header"
                      >
                        mtsamis
                      </h5>
                      <div
                        class="card-body"
                      >
                        <h5
                          class="card-title"
                        >
                          Thu Jul 14 2016
                        </h5>
                        <a
                          class="navbar-link text-decoration-none"
                          href="/questions/6ni6ok3ym7mf1p33lnez"
                        >
                          <div
                            class="d-grid gap-2"
                          >
                            <button
                              class="btn btn-info"
                              type="button"
                            >
                              Show
                            </button>
                          </div>
                        </a>
                      </div>
                    </div>
                  </div>
                  <div
                    class="col-4"
                  >
                    <div
                      class="card mb-2"
                    >
                      <h5
                        class="card-header"
                      >
                        sarahedo
                      </h5>
                      <div
                        class="card-body"
                      >
                        <h5
                          class="card-title"
                        >
                          Wed Jun 29 2016
                        </h5>
                        <a
                          class="navbar-link text-decoration-none"
                          href="/questions/8xf0y6ziyjabvozdd253nd"
                        >
                          <div
                            class="d-grid gap-2"
                          >
                            <button
                              class="btn btn-info"
                              type="button"
                            >
                              Show
                            </button>
                          </div>
                        </a>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
